File: src/test/java/com/hileco/cortex/constraints/SolverTest.java
package com.hileco.cortex.constraints;

import com.hileco.cortex.documentation.Documentation;
import com.hileco.cortex.instructions.conditions.LESS_THAN;
import com.hileco.cortex.instructions.io.LOAD;
import com.hileco.cortex.instructions.math.ADD;
import com.hileco.cortex.instructions.math.MODULO;
import com.hileco.cortex.instructions.stack.PUSH;
import org.junit.Assert;
import org.junit.Test;

import java.math.BigInteger;
import java.util.List;

import static com.hileco.cortex.vm.ProgramStoreZone.CALL_DATA;

public class SolverTest {

    @Test
    public void testSolve() {
        var instructions = List.of(
                new PUSH(BigInteger.valueOf(10L).toByteArray()),
                new PUSH(BigInteger.valueOf(0xffffffL).toByteArray()),
                new PUSH(BigInteger.valueOf(10L).toByteArray()),
                new PUSH(BigInteger.valueOf(0L).toByteArray()),
                new LOAD(CALL_DATA),
                new ADD(),
                new MODULO(),
                new LESS_THAN()
        );
        var expressionGenerator = new ExpressionGenerator();
        instructions.forEach(expressionGenerator::addInstruction);
        var solver = new Solver();
        var solution = solver.solve(expressionGenerator.getCurrentExpression());
        var onlyValue = solution.getPossibleValues().values().iterator().next();
        Documentation.of(Solver.class.getSimpleName())
                .headingParagraph(Solver.class.getSimpleName())
                .paragraph("Example program:").source(instructions)
                .paragraph("Resulting expression:").source(expressionGenerator.getCurrentExpression())
                .paragraph("Suggested solution for expression to be true:").source(solution);
        Assert.assertTrue(solution.isSolvable());
        Assert.assertTrue((10L + onlyValue) % 0xffffffL < 10);
    }
}

File: src/test/java/com/hileco/cortex/constraints/SolverTest.kt
package com.hileco.cortex.constraints

import com.hileco.cortex.documentation.Documentation
import com.hileco.cortex.instructions.conditions.EQUALS
import com.hileco.cortex.instructions.conditions.LESS_THAN
import com.hileco.cortex.instructions.io.LOAD
import com.hileco.cortex.instructions.math.ADD
import com.hileco.cortex.instructions.math.HASH
import com.hileco.cortex.instructions.math.MODULO
import com.hileco.cortex.instructions.stack.PUSH
import com.hileco.cortex.vm.ProgramStoreZone.CALL_DATA
import org.junit.Assert
import org.junit.Test

class SolverTest {
    @Test
    fun testSolve() {
        val instructions = listOf(
                PUSH(10),
                PUSH(0xffffff),
                PUSH(10),
                PUSH(0),
                LOAD(CALL_DATA),
                ADD(),
                MODULO(),
                LESS_THAN()
        )
        val expressionGenerator = ExpressionGenerator()
        instructions.forEach { expressionGenerator.addInstruction(it) }
        val solver = Solver()
        val solution = solver.solve(expressionGenerator.currentExpression)
        val onlyValue = solution.possibleValues.values.first()
        Documentation.of(Solver::class.simpleName!!)
                .headingParagraph(Solver::class.simpleName!!)
                .paragraph("Example program:").source(instructions)
                .paragraph("Resulting expression:").source(expressionGenerator.currentExpression)
                .paragraph("Suggested solution for expression to be true:").source(solution)
        Assert.assertTrue(solution.isSolvable)
        Assert.assertTrue((10L + onlyValue) % 0xffffffL < 10)
    }

    @Test
    fun testSolveUninterpretedFunction() {
        val instructions = listOf(
                PUSH(0),
                LOAD(CALL_DATA),
                HASH("SHA-256"),
                PUSH(10),
                HASH("SHA-256"),
                EQUALS()
        )
        val expressionGenerator = ExpressionGenerator()
        instructions.forEach { expressionGenerator.addInstruction(it) }
        val solver = Solver()
        val solution = solver.solve(expressionGenerator.currentExpression)
        Assert.assertTrue(solution.isSolvable)
    }
}